ghcide/src/Development/IDE/GHC/ExactPrint.hs

```diff
@@ -54,19 +61,30 @@ import Control.Arrow
 
 ------------------------------------------------------------------------------
 
+data GetAnnotatedParsedSource = GetAnnotatedParsedSource
+  deriving (Eq, Show, Typeable, GHC.Generic)
+
+instance Hashable GetAnnotatedParsedSource
+instance NFData GetAnnotatedParsedSource
+instance Binary GetAnnotatedParsedSource
+type instance RuleResult GetAnnotatedParsedSource = Annotated ParsedSource
+
 -- | Get the latest version of the annotated parse source.
-useAnnotatedSource ::
-    String ->
-    IdeState ->
-    NormalizedFilePath ->
-    IO (Maybe (Annotated ParsedSource))
-useAnnotatedSource herald state nfp =
-    fmap annotateParsedSource
-        <$> runAction herald state (use GetParsedModule nfp)
+getAnnotatedParsedSourceRule :: Rules ()
+getAnnotatedParsedSourceRule = define $ \GetAnnotatedParsedSource nfp -> do
+  pm <- use GetParsedModule nfp
+  return ([], fmap annotateParsedSource pm)
```

Comment:
This seems a fairly redundant rule, given its just a `get` over a `ParsedModule`. Why not just declare a function that secretly does the `need` on the underlying ParsedModule? Why put it on the graph at all?

Reply:
`getAnnotatedParsedSource` used to be a function, this PR caches the result in the Shake graph. This is because `annotatedParsedSource` is an expensive operation and the result is used by several plugins now: 

- splices
- tactics
- code actions

I checked with @alanz and he confirmed that this is worth caching.

Reply:
It looks like just a projection, didn't realise it was expensive - makes sense then!